### PR TITLE
Don't load genes with readthrough canonical transcripts

### PR DESCRIPTION
**Pull request by ipilizota**
## Description

When readthrough transcripts are processed by Compara gene tree pipeline, they get misannotated as paralogs of genes from which they were formed via splicing. Furthermore, the corresponding erroneous duplication nodes in the gene tree can mess up many other orthology and paralogy relationships.

We could avoid that by filtering out genes whose canonical transcript is a readthrough transcript. It seems convenient to simply not load such genes in the `LoadMembers` pipeline. 

**Related JIRA tickets:**
- ENSCOMPARASW-4465

## Overview of changes
Added a piece of code which discards genes whose canonical transcript is a readthrough transcript.

## Testing
* Full `LoadMembers pipeline run on 106 dataset: `http://guihive.ebi.ac.uk:8080/versions/96/?driver=mysql&username=ensadmin&host=mysql-ens-compara-prod-1&port=4485&dbname=ivana_vertebrates_load_members_4jan_106&passwd=xxxxx

* `SELECT COUNT(*) FROM gene_member WHERE genome_db_id = 150;`
In `twalsh_vertebrates_load_members_106`: 69326
In `ivana_vertebrates_load_members_4jan_106`: 68345
Difference: 981

* Query `homo_sapiens_core_106_38` ( `attrib_type_id=554` -> canonical transcript, `attrib_type_id=218` -> readhrough)
```
mysql-ens-vertannot-staging homo_sapiens_core_106_38

SELECT COUNT(*) FROM gene g, transcript t, transcript_attrib ta WHERE g.gene_id=t.gene_id AND t.transcript_id=ta.transcript_id AND ta.attrib_type_id=554 AND t.transcript_id IN (SELECT transcript_id FROM transcript_attrib WHERE attrib_type_id=218);
+----------+
| COUNT(*) |
+----------+
|      981 |
+----------+
```

* [Didn't redo this in Jan; I did it only before I made the PR] In addition, I checked messages for the job loading human core in the pipeline db: `msg` table. Numbers in messages `"Discarded XX genes because they have a readthrough canonical transcipt"` sum up to 981.

### CODE REVIEW

**Review by JAlvarezJarreta — CHANGES_REQUESTED** [review]
I like how you have avoided the usage of an internal ID that may change in the future :+1: 
To make the code even better, I would put all the new bit under a conditional parameter, such as `keep_canonical_transcripts`, so these can still be loaded if required (but discard them by default).

File: modules/Bio/EnsEMBL/Compara/RunnableDB/LoadMembers.pm [hunk 1]
```diff
@@ -223,6 +223,25 @@ sub loadMembersFromCoreSlices {
         $self->warning("Discarded ".($n1-scalar(@relevant_genes))." genes because they span the end of circular chromosomes");
     }
 
+    # Discard genes whose canonical transcripts are readthrough transcripts
+    my $n2 = scalar(@relevant_genes);
+    my @readthrough_genes;
+    my $attrib_value = 'readthrough';
+    foreach my $gene (@relevant_genes){
+        my $can_transcript = $gene->canonical_transcript();
+        my @tr_attribs = @{$can_transcript->get_all_Attributes()};
```
[the remark this answers is not in the document]
Comment (dthybert) [reply modules/Bio/EnsEMBL/Compara/RunnableDB/LoadMembers.pm:232]: I agree I think it is always better to have an explicit variable name.

File: modules/Bio/EnsEMBL/Compara/RunnableDB/LoadMembers.pm [hunk 2]
```diff
@@ -223,6 +223,25 @@ sub loadMembersFromCoreSlices {
         $self->warning("Discarded ".($n1-scalar(@relevant_genes))." genes because they span the end of circular chromosomes");
     }
 
+    # Discard genes whose canonical transcripts are readthrough transcripts
+    my $n2 = scalar(@relevant_genes);
+    my @readthrough_genes;
+    my $attrib_value = 'readthrough';
```
[the remark this answers is not in the document]
Comment (dthybert) [reply modules/Bio/EnsEMBL/Compara/RunnableDB/LoadMembers.pm:229]: Indeed this would make the code more readable.

**Comment by ipilizota** [comment]
Thank you very much for your feedback @JAlvarezJarreta. I'm addressing it now.

> To make the code even better, I would put all the new bit under a conditional parameter, such as `keep_canonical_transcripts`, so these can still be loaded if required (but discard them by default).

I see your point. The current code satisfies the **Definition of done** in the ticket and it looks like we are not planning to load genes whose canonical transcripts are readthrough transcripts. Shall we put your proposal into a new ticket? Do we have to do it at all?

**Comment by dthybert** [comment]
> Thank you very much for your feedback @JAlvarezJarreta. I'm addressing it now.
> 
> > To make the code even better, I would put all the new bit under a conditional parameter, such as `keep_canonical_transcripts`, so these can still be loaded if required (but discard them by default).
> 
> I see your point. The current code satisfies the **Definition of done** in the ticket and it looks like we are not planning to load genes whose canonical transcripts are readthrough transcripts. Shall we put your proposal into a new ticket? Do we have to do it at all?

Jorge comment  would be useful only if we want to do something with the  readthrough transcript (i.e gene tree for readthrough or anything else).At themoment we do not have use cases that would use aa readthrough transcript.  Even if we had it I doubt that we would want to load the readthrough transcript without a  readthrough tag to distinguish them from other coding genes . So this evolution would be a bit more complex and would need more thinking. I would keep this as it is for now and when we will face the need to load readthough we will look into how we do things.

**Comment by ipilizota** [comment]
All comments addressed. I ran the whole pipeline again and updated PR description accordingly.

**Review by JAlvarezJarreta — APPROVED** [review]

**Review by dthybert — APPROVED** [review]

**Pull request merged**